Use KubernetesHook to create api client in KubernetesPodOperator

dstandish opened the pull request:
Reopened and ready for review.

TLDR: add support for k8s hook in KPO; use it always (even when no conn id); continue to consider the core k8s settings that KPO already takes into account but emit deprecation warning about them.

This one was pretty tricky.

Here are some notes:

KPO takes into account a few settings from core airflow cfg (e.g. verify ssl, tcp keepalive, context, config file, and in_cluster).  So if we're using the hook to generate the client, somehow the hook has to take these settings into account. But we don't want the hook to consider these settings in general.  

So we need to read the settings in KPO and somehow pass them to the hook.  We can't do it through init params, because that would interfere with migration to an airflow connection.  So what I did was establish some "private" attribtues that could be patched from KPO when the setting is something other than the default.  Then the hook will consider these patched params only when it's not using an airflow connection or hook params.

The other tricky part was we want to minimize the warnings.  So we need to try and warn only when the configuration taken from core settings actually ends up being used.  E.g. `[kubernetes] > in_cluster=True` is the default value so we only want to warn when it's been set to False, and only when it is actually considered (so that when the user adds an airflow connection, the warning will go away).

Lastly one thing to note (smaller thing), with two settings, when adding them to the hook, I changed the wording.  Core airflow setting `verify_ssl` is True by default.  When adding this to the hook, I add it as `disable_verify_ssl`.  And the reason for this is, `False` seems like a safer default value.  So the default would be `disable_verify_ssl=False`.  E.g. if you did `if extra.get('disable_verify_ssl'): disable` that is safe when the key is not there.  But if you do `if not extra.get('verify_ssl'): disable` well this will disable SSL when the value is not there.  To make this work you'd have to do `if extra.get('verify_ssl') is False`.  The other thing

dstandish committed:
Use KubernetesHook to create api client in KubernetesPodOperator

BasPH commented on airflow/providers/cncf/kubernetes/CHANGELOG.rst:
```diff
@@ -103,6 +103,12 @@ Other changes in ``pod_launcher.py``:
+Features
+~~~~~~~~
+
+* Can now derive kubernetes creds using airflow connection, through KubernetesHook
```
```suggestion
* Can now derive Kubernetes credentials from an Airflow connection through the KubernetesHook
```

BasPH reviewed (commented):
> It doesn't seem that KubernetesHook lookss at these settings. Do you think we need to implement either / or logic here? That is, only use K8s hook if an airflow conn id is provided? And otherwise, continue to use get_kube_client?

Yea that seems reasonable to me

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -348,15 +351,17 @@ def launcher(self) -> pod_launcher.PodLauncher:
-    def client(self) -> CoreV1Api:
-        # todo: use airflow Connection / hook to authenticate to the cluster
-        kwargs: Dict[str, Any] = dict(
+    def k8s_hook(self):
```
```suggestion
    def hook(self):
```

nit, for consistency with other operators?

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -348,15 +351,17 @@ def launcher(self) -> pod_launcher.PodLauncher:
+
+    @cached_property
+    def client(self) -> CoreV1Api:
+        return self.k8s_hook.core_v1_client
```
```suggestion
        return self.hook.core_v1_client
```

jedcunningham reviewed (commented):
Probably makes sense to do the "or" logic with `get_kube_client`. I do wonder if we should emit a deprecation warning when a connection isn't used?

dstandish committed:
fixup! Use KubernetesHook to create api client in KubernetesPodOperator

dstandish committed:
Only apply in_cluster from core config if False

Since if hook not otherwise configured, we'll end up trying in_cluster anyway,
we don't really need to take into account a True value (the default in core).
We probably don't need to consider False either, but probably best to warn users
that consideration of their non-default config is going to be removed.

dstandish committed:
add tests for core settings patch on KPO

dstandish committed:
add tests for patch logic

dstandish committed:
remove variable

dstandish committed:
spelling

dstandish committed:
fix tests

dstandish committed:
fix KPO tests

dstandish committed:
try fixing kube pod tests

dstandish committed:
fix mypy

dstandish committed:
add docstring params

dstandish committed:
fix "system" tests

dstandish committed:
fix tests

dstandish committed:
fix system test

dstandish committed:
fix backcompat system test

jedcunningham commented on airflow/providers/cncf/kubernetes/CHANGELOG.rst:
```diff
@@ -19,6 +19,19 @@
+
+Previously, KubernetesPodOperator relied on core Airflow configuration (namely setting for kubernetes executor) for certain settings used in client generation.  Now KubernetesPodOperator uses KubernetesHook, and the consideration of core k8s settings is officially deprecated.
+
+If you are using the Airflow configuration settings (e.g. as opposed to operator params) to configure the kubernetes client, then prior to the next major release you will need to add an Airflow connection and set your KPO processes to use that connection.
```
```suggestion
If you are using the Airflow configuration settings (e.g. as opposed to operator params) to configure the kubernetes client, then prior to the next major release you will need to add an Airflow connection and set your KPO tasks to use that connection.
```

nit

jedcunningham commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -108,37 +121,77 @@ def __init__(
+            f"\nApplying core Airflow settings from section [kubernetes] with the following keys:"
+            f"{settings_list_str}\n"
+            "In a future release, KubernetesPodOperator will no longer consider core\n"
+            "airflow settings; define an Airflow connection instead.",
```
```suggestion
            "Airflow settings; define an Airflow connection instead.",
```

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -561,6 +569,39 @@ def dry_run(self) -> None:
+        """
+
+        # default for enable_tcp_keepalive is True; patch if False
+        if conf.getboolean('kubernetes', 'enable_tcp_keepalive', fallback=None) is False:
```
I'm curious, why do we need `fallback=None`?

dstandish replied:
if the value is missing from the config and you don't have a fallback, it would throw an error.
as to why None, well, just a reasonable default that means "not set".  Only when it's explicitly set to False would we want to set the hook to disable TCP keepalive.

jedcunningham replied:
For those following along at home, _some_ of these fallbacks had no effect (like this one) as conf falls back to the defaults first. However, some of the others have no default ever, and are necessary.

jedcunningham commented on kubernetes_tests/test_kubernetes_pod_operator.py:
```diff
@@ -582,10 +586,12 @@ def test_xcom_push(self, xcom_push):
-    @mock.patch("airflow.providers.cncf.kubernetes.utils.pod_manager.PodManager.await_pod_completion")
-    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
-    def test_envs_from_secrets(self, mock_client, await_pod_completion_mock, create_pod):
+    @mock.patch(f"{POD_MANAGER_CLASS}.create_pod")
```
👍

jedcunningham commented on tests/providers/cncf/kubernetes/hooks/test_kubernetes.py:
```diff
@@ -131,6 +136,76 @@ def test_get_default_client(
+        if disable_called:
+            assert mock_disable.called
+        else:
+            assert not mock_disable.called
```
```suggestion
        assert mock_disable.called is disabled_called
```

nit

dstandish replied:
🤯 mind blown

dstandish committed:
Update airflow/providers/cncf/kubernetes/CHANGELOG.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update airflow/providers/cncf/kubernetes/hooks/kubernetes.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update tests/providers/cncf/kubernetes/hooks/test_kubernetes.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
remove fallbacks that have no effect

dstandish committed:
simplify assert in `test_disable_tcp_keepalive`

jedcunningham commented on tests/providers/cncf/kubernetes/hooks/test_kubernetes.py:
```diff
@@ -131,6 +136,73 @@ def test_get_default_client(
+        """
+        kubernetes_hook = KubernetesHook(conn_id=conn_id, disable_verify_ssl=disable_verify_ssl)
+        api_conn = kubernetes_hook.get_conn()
+        assert mock_disable.called is disabled_called
```
```suggestion
        assert mock_disable.called is disable_called
```

Oops! Sorry for my typo.

jedcunningham committed:
Update tests/providers/cncf/kubernetes/hooks/test_kubernetes.py

dstandish committed:
remove dupe param
